### ISSUE TITLE
backup: Parallelize processing of files

### ISSUE THREAD

**Issue by deajan** [rustic-rs/rustic#554]
Hello,

Still working on my backup benchmarks, my first round saw rustic being the slowest solution so far.
My workload is to backup the linux kernel sources multiple times, and change tag between backups, so only some files get modified.
This workload runs 4 times, then a restore operation is done.

rustic tends to be 3-4x slower than restic on backup operations.
restore operation took about the same time restic took.

I've run rustic without any special parameters.
The only thing I can think of is that since the kernel sources are about 70k files and rustic does not parallelize file analyze, there might be the culprit.

Any wild guesses of what I can do to properly benchmark rustic ?
My test env: Xeon-1260L 64GB RAM with RHEL9 and SATA SSD disks.

**Comment by aawsome** [comment]
Can you also post some timings you did measure (best would be some output of rustic)? Are you talking about seconds, minutes or hours of backup time?

About your results: It depends a bit about how you define performance. 
Your test system has 8 threads. rustic so far doesn't process files in parallel, whereas restic does. So, a time factor up to 7-8 for your test-setting with many small files would be quite reasonable.

Did you also compare the total resources (total CPU time, memory, I/O) used? I/O should be very similar, of course.
Did you make a comparison using just one big file? 

Actually, the point is about what you want to tune when you do performance tuning. For all my current use cases rustic is fast enough to finish it's backup as desired (usually only a couple of seconds, as I don't have many changed files). Here it doesn't matter much if I could parallelize things more which would reduce the backup time to fewer seconds but still uses the same total CPU time - the load on the system would be equal, just more peaky. And if backups finish in time, there is no need to improve this.

This changes if you have a lot of data changing and you are getting into trouble finishing your backups in time.
But as rustic supports basically all operations concurrently on a repository, you won't run into trouble that soon, IMO.
Would be nice, however, to know at which size of changed data we reach the point where parallelization is cruical.

**Comment by aawsome** [comment]
@deajan I was able to spend 2h on profiling this evening and found some general optimizations as well as starting points for better parallelization.
This is still all WIP and there are quite some open points for improvement. However, I would really appreciate if you could test the current version of #556 with your benchmark - at least the resources used should be now way below restic.
BTW: the binaries from the CI builds are available [here](https://github.com/rustic-rs/rustic/actions/runs/4736820597?pr=556) (scroll down to the bottom of the page)

**Comment by aawsome** [comment]
I did a bit more testing and performance measurement on #556 and merged it. So, binaries are now also available as beta builds.
I'm still not satisfied with the parallelization result of #556 - restic seems to be able to parallelize even more - but I wasn't able so far to find the bottleneck which stops rustic from doing even more parallelization. But anyhow it is already a big improvement (around factor 5 in my tests) and I'm missing time to dig deeper. So I merged it - but if anyone is willing to jump into this topic, give me a call and I'll happily assist to find even better solutions!
